src/test/regress/expected/distributed_functions.out

```diff
@@ -52,6 +52,190 @@ SELECT * FROM run_command_on_workers('SELECT function_tests.dup(42);') ORDER BY
+ 
+(1 row)
+
+-- valid distribution with distribution_arg_name -- case insensitive
+SELECT create_distributed_function('add_with_param_names(int, int)', distribution_arg_name:='VaL1');
+ create_distributed_function 
+-----------------------------
+ 
+(1 row)
+
+-- valid distribution with distribution_arg_index
+SELECT create_distributed_function('add_with_param_names(int, int)', distribution_arg_index:=1);
+ create_distributed_function 
+-----------------------------
+ 
+(1 row)
+
+SELECT create_distributed_function('add_without_param_names(int, int)', distribution_arg_index:=1);
+ create_distributed_function 
+-----------------------------
+ 
+(1 row)
+
+-- a function cannot be colocated with a table that is not "streaming" replicated 
+SET citus.shard_replication_factor TO 2;
+CREATE TABLE replicated_table_func_test (a int);
+SET citus.replication_model TO "statement";
+SELECT create_distributed_table('replicated_table_func_test', 'a');
+ create_distributed_table 
+--------------------------
+ 
+(1 row)
+
+SELECT create_distributed_function('add_with_param_names(int, int)', distribution_arg_index:=1, colocate_with:='replicated_table_func_test');
+ERROR:  cannot colocate function "add_with_param_names" and table "replicated_table_func_test"
+DETAIL:  Citus currently only supports colocating function with distributed tables that are created using streaming replication model.
+HINT:  When distributing tables make sure that "citus.replication_model" is set to "streaming"
+-- a function cannot be colocated with a different distribution argument type
+SET citus.shard_replication_factor TO 1;
+CREATE TABLE replicated_table_func_test_2 (a bigint);
+SET citus.replication_model TO "streaming";
+SELECT create_distributed_table('replicated_table_func_test_2', 'a');
+ create_distributed_table 
+--------------------------
+ 
+(1 row)
+
+SELECT create_distributed_function('add_with_param_names(int, int)', distribution_arg_index:=1, colocate_with:='replicated_table_func_test_2');
+ERROR:  cannot colocate function "replicated_table_func_test_2" and table "add_with_param_names" because distribution column types don't match
+-- a function cannot be colocated with a local table
+CREATE TABLE replicated_table_func_test_3 (a bigint);
+SELECT create_distributed_function('add_with_param_names(int, int)', distribution_arg_index:=1, colocate_with:='replicated_table_func_test_3');
+ERROR:  relation replicated_table_func_test_3 is not distributed
+-- a function cannot be colocated with a reference table
+SELECT create_reference_table('replicated_table_func_test_3');
+ create_reference_table 
+------------------------
+ 
+(1 row)
+
+SELECT create_distributed_function('add_with_param_names(int, int)', distribution_arg_index:=1, colocate_with:='replicated_table_func_test_3');
+ERROR:  cannot colocate function "add_with_param_names" and table "replicated_table_func_test_3" because colocate_with option is only supported for hash distributed tables.
```

Comment:
What happens if we allow this?

Reply:
I think there are two things to consider:

- Hash tables with rep > 1, append/range tables: We do not sync metadata for such distributed tables anyway. So, it doesn't help much to allow, and could be confusing.

- For reference tables: I think, it makes sense to distribute `reference function`. But, to keep the scope smaller and also follow our naming conventions (e.g., we should probably have `create_reference_function()`). Do you think should we go ahead and implement it?

Reply:
Reference function sounds pretty odd. Anyway, it also complicates the CALL logic a bit so let's maybe punt on this.